presto-main/src/main/java/com/facebook/presto/catalogserver/CatalogServer.java

```diff
@@ -55,103 +71,198 @@ public CatalogServer(MetadataManager metadataProvider, SessionPropertyManager se
-            return writeValueAsString(schemaNames, objectMapper);
-        }
-        return EMPTY_STRING;
+        CacheKey cacheKey = new CacheKey(session, catalogName);
+        boolean isCacheHit = isCacheHit(cacheKey, listSchemaNamesCache);
+        List<String> schemaNames = listSchemaNamesCache.getUnchecked(cacheKey);
+        return schemaNames.isEmpty()
+                ? new CatalogServerClient.MetadataEntry<>(EMPTY_STRING, isCacheHit)
+                : new CatalogServerClient.MetadataEntry<>(writeValueAsString(schemaNames, objectMapper), isCacheHit);
     }
 
     @ThriftMethod
-    public String getTableHandle(TransactionInfo transactionInfo, SessionRepresentation session, QualifiedObjectName table)
+    public CatalogServerClient.MetadataEntry<String> getTableHandle(TransactionInfo transactionInfo, SessionRepresentation session, QualifiedObjectName table)
     {
         transactionManager.tryRegisterTransaction(transactionInfo);
-        Optional<TableHandle> tableHandle = metadataProvider.getTableHandle(session.toSession(sessionPropertyManager), table);
-        return tableHandle.map(handle -> writeValueAsString(handle, objectMapper))
-                .orElse(EMPTY_STRING);
+        CacheKey cacheKey = new CacheKey(session, table);
+        boolean isCacheHit = isCacheHit(cacheKey, getTableHandleCache);
+        Optional<TableHandle> tableHandle = getTableHandleCache.getUnchecked(cacheKey);
+        if (!tableHandle.isPresent()) {
+            getTableHandleCache.refresh(cacheKey);
+            tableHandle = getTableHandleCache.getUnchecked(cacheKey);
+        }
+        return tableHandle.isPresent()
+                ? new CatalogServerClient.MetadataEntry<>(writeValueAsString(tableHandle.get(), objectMapper), isCacheHit)
+                : new CatalogServerClient.MetadataEntry<>(EMPTY_STRING, isCacheHit);
     }
 
     @ThriftMethod
-    public String listTables(TransactionInfo transactionInfo, SessionRepresentation session, QualifiedTablePrefix prefix)
+    public CatalogServerClient.MetadataEntry<String> listTables(TransactionInfo transactionInfo, SessionRepresentation session, QualifiedTablePrefix prefix)
     {
         transactionManager.tryRegisterTransaction(transactionInfo);
-        List<QualifiedObjectName> tableList = metadataProvider.listTables(session.toSession(sessionPropertyManager), prefix);
-        if (!tableList.isEmpty()) {
-            return writeValueAsString(tableList, objectMapper);
-        }
-        return EMPTY_STRING;
+        CacheKey cacheKey = new CacheKey(session, prefix);
+        boolean isCacheHit = isCacheHit(cacheKey, listTablesCache);
+        List<QualifiedObjectName> tableList = listTablesCache.getUnchecked(cacheKey);
+        return tableList.isEmpty()
+                ? new CatalogServerClient.MetadataEntry<>(EMPTY_STRING, isCacheHit)
+                : new CatalogServerClient.MetadataEntry<>(writeValueAsString(tableList, objectMapper), isCacheHit);
```

Comment:
`.map(...).orElseGet(...)`

Reply:
Think you mistakened this for an optional?